src/agent/metrics.rs

```diff
@@ -0,0 +1,376 @@
+                    if let Some(datetime) =
+                        NaiveDateTime::from_timestamp_opt(local_data.timestamp, 0)
+                    {
+                        datetime.format("%Y-%m-%d %H:%M:%S").to_string()
+                    } else {
+                        format!("Invalid timestamp {}", local_data.timestamp)
+                    }
+                } else {
+                    "no data".to_string()
+                };
+
+                let row_snippet = html! {
+                            <tr>
+                                <td>{text!(symbol.clone())}</td>
+                                <td>{text!(data.product.to_string())}</td>
+                <td>{text!(price_pubkey.to_string())}</td>
+                <td>{text!(price_string)}</td>
+                <td>{text!(last_publish_string)}</td>
+                <td>{text!(last_local_update_string)}</td>
+                            </tr>
+                            };
+                rows.push(row_snippet);
+            }
+        }
+
+        let title_string = concat!("Pyth Agent Dashboard - ", env!("CARGO_PKG_VERSION"));
+        let res_html: DOMTree<String> = html! {
+        <html>
+            <head>
+            <title>{text!(title_string)}</title>
+        <style>
+            """
+table {
+  width: 100%;
+  border-collapse: collapse;
+}
+table, th, td {
+  border: 1px solid;
+}
+"""
+        </style>
+            </head>
+            <body>
+            <h1>{text!(title_string)}</h1>
+        {text!("Uptime: {}", uptime_string)}
+            <h2>"State Overview"</h2>
+            <table>
+            <tr>
+                <th>"Symbol"</th>
+                <th>"Product ID"</th>
+                <th>"Price ID"</th>
+                <th>"Last Published Price"</th>
+        <th>"Last Publish Time"</th>
+        <th>"Last Local Update Time"</th>
+            </tr>
+            { rows }
+        </table>
+            </body>
+        </html>
+        };
+        Ok(res_html.to_string())
+    }
+}
+
+pub struct DashboardSymbolView {
+    product: Pubkey,
+    prices:  BTreeMap<Pubkey, DashboardPriceView>,
+}
+
+pub struct DashboardPriceView {
+    local_data:      Option<PriceInfo>,
+    global_data:     Option<PriceAccount>,
+    global_metadata: Option<PriceAccountMetadata>,
+}
+
+/// Turn global/local store state into a single per-symbol view.
+///
+/// The dashboard data comes from three sources - the global store
+/// (observed on-chain state) data, global store metadata and local
+/// store data (local state possibly not yet committed to the oracle
+/// contract).
+///
+/// The view is indexed by human-readable symbol name or a stringified
+/// public key if symbol name can't be found.
+pub fn build_dashboard_data(
+    mut local_data: HashMap<PriceIdentifier, PriceInfo>,
+    mut global_data: AllAccountsData,
+    mut global_metadata: AllAccountsMetadata,
+    logger: &Logger,
+) -> BTreeMap<String, DashboardSymbolView> {
+    let mut ret = BTreeMap::new();
```

Comment:
Is there any reason that it's a map?

Reply:
BTree* family containers automatically sort their keys as opposed to the regular Hash* implementations. The goal was to get a sorted view every time

Reply:
The primary reason I used a map was deduplication. Come to think of it, I should probably detect overwrites on existing entries